quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/TruststoreBuilder.java

```diff
@@ -0,0 +1,190 @@
+
+        // finally update the system properties
+        System.setProperty(SYSTEM_TRUSTSTORE_KEY, file.getAbsolutePath());
+        System.setProperty(SYSTEM_TRUSTSTORE_TYPE_KEY, PKCS12);
+        System.setProperty(SYSTEM_TRUSTSTORE_PASSWORD_KEY, CHANGEIT);
+    }
+
+    static File saveTruststore(KeyStore truststore, String dataDir, char[] password) {
+        File file = new File(dataDir, "keycloak-truststore.p12");
+        file.getParentFile().mkdirs();
+        try (FileOutputStream fos = new FileOutputStream(file)) {
+            System.setProperty("keystore.pkcs12.certProtectionAlgorithm", "NONE");
+            truststore.store(fos, password);
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to save truststore: " + file.getAbsolutePath(), e);
+        }
+        return file;
+    }
+
+    static KeyStore createMergedTruststore(String[] truststores, boolean trustStoreIncludeDefault) {
+        KeyStore truststore = createPkcs12KeyStore();
+
+        if (trustStoreIncludeDefault) {
+            includeDefaultTruststore(truststore);
+        }
+
+        for (String file : truststores) {
+            if (file.endsWith(".p12") || file.endsWith(".pfx")) {
+                mergeTrustStore(truststore, file, loadStore(file, PKCS12, null));
+            } else {
+                mergePemFile(truststore, file);
+            }
+        }
+        return truststore;
+    }
+
+    static KeyStore createPkcs12KeyStore() {
+        try {
+            KeyStore truststore = KeyStore.getInstance(PKCS12);
+            truststore.load(null, null);
+            return truststore;
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to initialize truststore: cannot create a PKCS12 keystore", e);
+        }
+    }
+
+    static void includeDefaultTruststore(KeyStore truststore) {
+        String originalTruststoreKey = SYSTEM_TRUSTSTORE_KEY + ".orig";
+        String originalTruststoreTypeKey = SYSTEM_TRUSTSTORE_TYPE_KEY + ".orig";
+        String originalTruststorePasswordKey = SYSTEM_TRUSTSTORE_PASSWORD_KEY + ".orig";
+
+        String trustStorePath = System.getProperty(originalTruststoreKey);
+        String type = PKCS12;
+        String password = null;
+        File defaultTrustStore = null;
+        if (trustStorePath == null) {
+            trustStorePath = System.getProperty(SYSTEM_TRUSTSTORE_KEY);
+            if (trustStorePath == null) {
+                // try jre locations
+                String securityDirectory = System.getProperty("java.home") + File.separator + "lib" + File.separator
+                        + "security";
+                File jssecacertsFile = new File(securityDirectory, "jssecacerts");
+                if (jssecacertsFile.exists() && jssecacertsFile.isFile()) {
+                    defaultTrustStore = jssecacertsFile;
+                } else {
+                    defaultTrustStore = new File(securityDirectory, "cacerts");
+                }
```

Comment:
This feels a bit fragile and too low-level as it feels it might even be JVM impl specific. But seems to be aligned with what others are doing, so probably ok.